ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java

```diff
@@ -176,20 +185,34 @@ private <T extends CreateAsSelect> ConfiguredStatement<?> injectForCreateAsSelec
       }
     }
 
+    throwIfRetentionPresentForTable(topicCleanUpPolicy, properties.getRetentionInMillis());
+
     final TopicProperties info
         = createTopic(topicPropertiesBuilder, topicCleanUpPolicy, additionalTopicConfigs);
 
     final T withTopic = (T) createAsSelect.copyWith(properties.withTopic(
         info.getTopicName(),
         info.getPartitions(),
-        info.getReplicas()
+        info.getReplicas(),
+        (Long) additionalTopicConfigs
+            .getOrDefault(TopicConfig.RETENTION_MS_CONFIG, info.getRetentionInMillis())
     ));
 
     final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
 
     return statement.withStatement(withTopicText, withTopic);
   }
 
+  private void throwIfRetentionPresentForTable(
+      final String topicCleanUpPolicy,
+      final Optional<Long> retentionInMillis
+  ) {
+    if (topicCleanUpPolicy.equals(TopicConfig.CLEANUP_POLICY_COMPACT)
+        && retentionInMillis.isPresent()) {
+      throw new KsqlException("Invalid config variable in the WITH clause: RETENTION_MS");
```

Comment:
```
Invalid config variable in the WITH clause: RETENTION_MS. Non-windowed tables do not support retention.
```
OR
```
Invalid config variable in the WITH clause: RETENTION_MS. This config cannot be applied when the cleanup.policy setting is set to compact.
```
@agavra will any of the above error messages help?

Reply:
I would go with the first option above, @bvarghese1. Some folks may not be aware of the relationship between ksql tables and Kafka compaction. Best to avoid increasing the conceptual scope.

Reply:
+1

Reply:
Done